### PR TITLE
Fix apply function

### LINKED ISSUE #9234: BUG: apply_function arg/kwarg handling
`raw/epochs/evoked.apply_function` all have an `*args, **kwargs` in their signatures. But they don't work the way I think they should / the docs are rather misleading. Preparation:

```py
import os
import mne
from scipy.ndimage import gaussian_filter

sample_data_folder = mne.datasets.sample.data_path()
sample_data_raw_file = os.path.join(sample_data_folder, 'MEG', 'sample',
                                    'sample_audvis_raw.fif')
raw = mne.io.read_raw_fif(sample_data_raw_file, verbose=False, preload=False)
raw.crop(tmax=5).load_data()
```

`gaussian_filter1d` requires a `sigma` parameter in addition to the data. This works as expected:
```py
raw.apply_function(gaussian_filter1d, sigma=4)
```
This also works:
```py
kwargs = dict(sigma=4)
raw.apply_function(gaussian_filter1d, **kwargs)
```

This next one is expected to fail **if you already know how argument unpacking works**, and does fail (`4.` is passed as `picks` param of `apply_function`, so you get `picks must be a list of int or list of str, got a data type of float64`):
```py
args = (4.,)
raw.apply_function(gaussian_filter1d, *args)
```
But note that if `args = (4,)` (int instead of float) you get a different error, that `gaussian_filter1d` is missing required argument `sigma`).  From the perspective of a new user, this is really bad: `*args` is in the docstring, so I pass in `*args` and it fails, in 2 different ways depending on what seems like a trivial difference in input.

This also fails (`gaussian_filter1d() got an unexpected keyword argument 'args'`) though might have been expected/tried by a new user:
```py
args = (4.,)
raw.apply_function(gaussian_filter1d, args=args)
```

This also fails due to python syntax restrictions, but might have been expected to work based on the docstring:
```py
raw.apply_function(gaussian_filter1d, verbose=None, *args)
```

So in sum we don't actually support passing `args` to the applied function.  Possible solutions:

1. change function signature from `*args, *kwargs` to `args, kwargs`, and accept only tuples for `args` and only dicts for `kwargs`
2. drop `*args` from the function signature, leave kwarg handling as it is

### PR DESCRIPTION
**Pull request by drammock**
closes #9234 

I've locally updated the test for `epochs.apply_function()` and I'm a bit surprised / wondering if there's another bug in there.  Marking as draft until I've figured it out one way or another.  It seems like when applied channel-wise it gets applied to an ndarray that is `(n_epochs, n_times)` and it's not clear to me that that's the right thing to do (maybe it should really be epoch-wise and channel-wise?)

cc @nordme and @vpeterson

### CODE REVIEW

**Review by agramfort — APPROVED** [review]

File: mne/io/tests/test_apply_function.py [hunk 1]
```diff
@@ -63,6 +63,3 @@ def test_apply_function_verbose():
         assert out is raw
         raw.apply_function(printer, verbose=True)
         assert sio.getvalue().count('\n') == n_chan
-
-
-run_tests_if_main()
```
Comment (larsoner) [inline mne/io/tests/test_apply_function.py:68]:
```
Warning: ./mne/io/tests/test_apply_function.py:10:1: F401 'mne.utils.run_tests_if_main' imported but unused
```

File: mne/tests/test_epochs.py [hunk 2]
```diff
@@ -3380,36 +3380,34 @@ def test_empty_constructor():
...
+
+    # apply_function to all channels at once
+    def fun(data):
+        """Reverse channel order without changing values."""
+        return np.eye(data.shape[1])[::-1] @ data
+
+    want = data_epochs[:, ::-1]
+    got = epochs.apply_function(fun, channel_wise=False).get_data()
+    assert np.array_equal(want, got)
+    assert got.shape == data_epochs.shape
```
Comment (drammock) [inline mne/tests/test_epochs.py:3400]:
```suggestion
```

Comment (drammock) [reply mne/tests/test_epochs.py:3400]: this is redundant (we already check shape using `_check_fun` when it runs, and `np.array_equal` also checks shape)

File: mne/tests/test_epochs.py [hunk 3]
```diff
@@ -3380,36 +3380,34 @@ def test_empty_constructor():
...
+
+    # apply_function channel-wise (to first 3 channels) by replacing with mean
+    picks = np.arange(3)
+    non_picks = np.arange(3, n_channels)
+
+    def fun(data):
+        return np.full_like(data, data.mean())
+
+    out = epochs.apply_function(fun, picks=picks, channel_wise=True)
+
```
Comment (drammock) [inline mne/tests/test_epochs.py:3410]:
```suggestion
```

File: mne/tests/test_epochs.py [hunk 4]
```diff
@@ -3380,36 +3380,32 @@ def test_empty_constructor():
...
-    assert check_picks and check_nopicks
+
+    # apply_function to all channels at once
+    def fun(data):
+        """Reverse channel order without changing values."""
+        return np.eye(data.shape[1])[::-1] @ data
+
+    want = data_epochs[:, ::-1]
+    got = epochs.apply_function(fun, channel_wise=False).get_data()
+    assert np.array_equal(want, got)
```
Comment (larsoner) [inline mne/tests/test_epochs.py:3399]: assert_array_equal is our standard way to do this (the failure traceback is better, etc.)

File: mne/tests/test_epochs.py [hunk 5]
```diff
@@ -3380,36 +3380,32 @@ def test_empty_constructor():
...
+    picks = np.arange(3)
+    non_picks = np.arange(3, n_channels)
+
+    def fun(data):
+        return np.full_like(data, data.mean())
+
+    out = epochs.apply_function(fun, picks=picks, channel_wise=True)
+    expected = epochs.get_data(picks).mean(axis=-1, keepdims=True)
+    assert np.all(out.get_data(picks) == expected)
+    assert np.array_equal(out.get_data(non_picks), epochs.get_data(non_picks))
```
Comment (larsoner) [inline mne/tests/test_epochs.py:3411]: Same for these

File: mne/tests/test_epochs.py (hunk 4 above)
Comment (drammock) [reply mne/tests/test_epochs.py:3399]:
```suggestion
    assert_array_equal(want, got)
```

File: mne/tests/test_epochs.py (hunk 5 above)
Comment (drammock) [reply mne/tests/test_epochs.py:3411]:
```suggestion
    assert_array_equal(out.get_data(non_picks), epochs.get_data(non_picks))
```

**Comment by drammock** [comment]
FYI for @vpeterson and @nordme: after a dev discussion on Discord, it was decided to change the behavior here.  What you implemented for `epochs.apply_function(..., channel_wise=True)` would pass an `n_epochs × n_times` array to the applied function.  What I've now implemented will pass only a 1-D time series (one channel within 1 epoch) to the applied function.

**Comment by larsoner** [comment]
Thanks @drammock !

**Pull request merged**